src/coprocessor/dag/rpn_expr/types/expr_eval.rs

```diff
@@ -745,27 +729,86 @@ mod tests {
         ];
 
         // FnFoo(col1, col0)
-        let rpn_nodes = vec![
-            RpnExpressionNode::ColumnRef { offset: 1 },
-            RpnExpressionNode::ColumnRef { offset: 0 },
-            RpnExpressionNode::FnCall {
-                func: Box::new(FnFoo),
-                field_type: {
-                    let mut ft = FieldType::new();
-                    ft.as_mut_accessor().set_tp(FieldTypeTp::LongLong);
-                    ft
-                },
+        let exp = RpnExpressionBuilder::new()
+            .push_column_ref(1)
+            .push_column_ref(0)
+            .push_fn_call(FnFoo, FieldTypeBuilder::new().tp(FieldTypeTp::LongLong))
+            .build();
+        let mut ctx = EvalContext::default();
+        let result = exp.eval(&mut ctx, 3, schema, &mut columns);
+        let val = result.unwrap();
+        assert!(val.is_vector());
+        assert_eq!(
+            val.vector_value().unwrap().as_int_slice(),
+            [Some(-2), Some(-17), Some(22)]
+        );
+        assert_eq!(val.field_type().tp(), FieldTypeTp::LongLong);
+    }
+
+    /// Binary function (arguments are both raw columns). The same column is referred multiple times
+    /// and it should be Ok.
+    #[test]
+    fn test_eval_binary_function_raw_column() {
+        /// FnFoo(v1, v2) performs v1 * v2.
+        #[derive(Debug, Clone, Copy)]
+        struct FnFoo;
+
+        impl_template_fn! { 2 arg @ FnFoo }
+
+        impl FnFoo {
+            fn call(
+                _ctx: &mut EvalContext,
+                _payload: RpnFnCallPayload<'_>,
+                v1: &Option<i64>,
+                v2: &Option<i64>,
+            ) -> Result<Option<i64>> {
+                Ok(Some(v1.unwrap() * v2.unwrap()))
+            }
+        }
+
+        let mut columns = LazyBatchColumnVec::from(vec![{
+            let mut col = LazyBatchColumn::raw_with_capacity(3);
+
+            let mut datum_raw = Vec::new();
+            DatumEncoder::encode(&mut datum_raw, &[Datum::I64(-5)], false).unwrap();
+            col.push_raw(&datum_raw);
+
+            let mut datum_raw = Vec::new();
+            DatumEncoder::encode(&mut datum_raw, &[Datum::I64(-7)], false).unwrap();
+            col.push_raw(&datum_raw);
+
+            let mut datum_raw = Vec::new();
+            DatumEncoder::encode(&mut datum_raw, &[Datum::I64(3)], false).unwrap();
+            col.push_raw(&datum_raw);
+
+            col
+        }]);
+
+        let schema = &[
+            {
+                let mut ft = FieldType::new();
```

Comment:
Why not use the builder?

Reply:
Replace existing code using FieldTypeBuilder should be another PR because this PR adds RpnExpressionBuilder.